// grpc interface for EVM environment (nodejs server & python client)

syntax = "proto3";

package evm;

// Interface exported by the server.
service EVM {

  rpc Reset(Status) returns (stream Status) {}
  rpc GetAccounts(Status) returns (stream Json) {}
  rpc Compile(Source) returns (Json) {}
  rpc Deploy(Json) returns (stream Address) {}
  rpc SendTx(SendTxData) returns (stream Json) {}
  
}

message Status {
  uint32 option = 1;
}

message Source {
  // plain string of contract source
  string text = 1;
  // contract name
  string name = 2;
}

message Json {
  // json string of solc output
  string data = 1;
}

message Address {
  // hex string of address, e.g. "0x1234..."
  string address = 1;
}

message SendTxData {
  // address of caller
  string fromAddr = 1;
  // address of callee
  string toAddr = 2;
  // hex string of sent value
  string value = 3;
  // hex string of payload
  string data = 4;
}